### PR TITLE
testnet sync fix - bad blocks on exception list (again... learning git, lol)

### PR DESCRIPTION
**Pull request by Quezacoatl1**

### CODE REVIEW

File: src/main.cpp [hunk 1]
```diff
@@ -3389,21 +3389,32 @@ bool CBlock::ConnectBlock(CTxDB& txdb, CBlockIndex* pindex, bool fJustCheck, boo
...
+                        {   uint256("129ae6779d620ec189f8e5148e205efca2dfe31d9f88004b918da3342157b7ff") //T407024
+                           ,uint256("c3f85818ba5290aaea1bcbd25b4e136f83acc93999942942bbb25aee2c655f7a") //T407068
+                           ,uint256("cf7f1316f92547f611852cf738fc7a4a643a2bb5b9290a33cd2f9425f44cc3f9") //T407099
+                           ,uint256("b47085beb075672c6f20d059633d0cad4dba9c5c20f1853d35455b75dc5d54a9") //T407117
+                           ,uint256("5a7d437d15bccc41ee8e39143e77960781f3dcf08697a888fa8c4af8a4965682") //T407161
+                           ,uint256("aeb3c24277ae1047bda548975a515e9d353d6e12a2952fb733da03f92438fb0f") //T407181
+                           ,uint256("fc584b18239f3e3ea78afbbd33af7c6a29bb518b8299f01c1ed4b52d19413d4f") //T407214
+                           ,uint256("d5441f7c35eb9ea1b786bbbed820b7f327504301ae70ef2ac3ca3cbc7106236b") //T479114
+                        };
+                        if( vSkipHashdStakeRewardSignCheck.count(pindex->GetBlockHash())==0 )
```
Comment (denravonska) [inline src/main.cpp]: We want to limit this check to testnet, right?

**Comment by denravonska** [comment]
Set your editor's tab width to 4 spaces ;)

**Review by tomasbrod — CHANGES_REQUESTED** [review]
Please change the skip condition to still execute the checks, but only LogPrintf warning instead of return error.
This way the diff should be smaller.
Please change the log message to something true, this is reward check not signature/hash check.

File: src/main.cpp (hunk 1 above)
Comment (Quezacoatl1) [reply src/main.cpp]: I agree, when I get the chance I will change the PR and add a check to the if.

**Comment by Quezacoatl1** [comment]
@tomasbrod will try to suit your needs

**Comment by Quezacoatl1** [comment]
@denravonska @tomasbrod 
I changed my tab width, the logic and the warning message. Sync from 0 is being tested. I hope I did not completely mess up/misunderstand the if statements as I am very very far from being certain about what I did :D Small steps...

File: src/main.cpp [hunk 2]
```diff
@@ -3389,19 +3389,30 @@ bool CBlock::ConnectBlock(CTxDB& txdb, CBlockIndex* pindex, bool fJustCheck, boo
...
+                           ,uint256("aeb3c24277ae1047bda548975a515e9d353d6e12a2952fb733da03f92438fb0f") //T407181
+                           ,uint256("fc584b18239f3e3ea78afbbd33af7c6a29bb518b8299f01c1ed4b52d19413d4f") //T407214
+                           ,uint256("d5441f7c35eb9ea1b786bbbed820b7f327504301ae70ef2ac3ca3cbc7106236b") //T479114
+                        };
                         if (dStakeReward > ((OUT_POR*1.25)+OUT_INTEREST+1+CoinToDouble(nFees)))
                         {
                             StructCPID st1 = GetLifetimeCPID(pindex->GetCPID(),"ConnectBlock()");
-                            GetProofOfStakeReward(nCoinAge, nFees, bb.cpid, true, 2, nTime,
-                                        pindex, "connectblock_researcher_doublecheck", OUT_POR, OUT_INTEREST, dAccrualAge, dMagnitudeUnit, dAvgMagnitude);
+                            GetProofOfStakeReward(nCoinAge, nFees, bb.cpid, true, 2, nTime, pindex, "connectblock_researcher_doublecheck", OUT_POR, OUT_INTEREST, dAccrualAge, dMagnitudeUnit, dAvgMagnitude);
```
Comment (tomasbrod) [inline src/main.cpp]: regression in readability

File: src/main.cpp [hunk 3]
```diff
@@ -3389,19 +3389,30 @@ bool CBlock::ConnectBlock(CTxDB& txdb, CBlockIndex* pindex, bool fJustCheck, boo
...
-                                    (double)bb.InterestSubsidy,(double)bb.ResearchSubsidy,dStakeReward,(double)OUT_POR,(double)OUT_INTEREST,bb.cpid.c_str()));
+                                if( vSkipHashdStakeRewardSignCheck.count(pindex->GetBlockHash())==0 || (vSkipHashdStakeRewardSignCheck.count(pindex->GetBlockHash())!=0 && !fTestNet))
+                                {
+                                    if (fDebug3) LogPrintf("ConnectBlockError[ResearchAge] : Researchers Reward Pays too much : Interest %f and Research %f and StakeReward %f, OUT_POR %f, with Out_Interest %f for CPID %s ",
+                                        (double)bb.InterestSubsidy,(double)bb.ResearchSubsidy,dStakeReward,(double)OUT_POR,(double)OUT_INTEREST,bb.cpid);
+                                        return DoS(10,error("ConnectBlock[ResearchAge] : Researchers Reward Pays too much : Interest %f and Research %f and StakeReward %f, OUT_POR %f, with Out_Interest %f for CPID %s ",
+                                            (double)bb.InterestSubsidy,(double)bb.ResearchSubsidy,dStakeReward,(double)OUT_POR,(double)OUT_INTEREST,bb.cpid.c_str()));
```
Comment (tomasbrod) [inline src/main.cpp:3414]: misleading indent

File: src/main.cpp [hunk 4]
```diff
@@ -3389,19 +3389,30 @@ bool CBlock::ConnectBlock(CTxDB& txdb, CBlockIndex* pindex, bool fJustCheck, boo
...
                             if (dStakeReward > ((OUT_POR*1.25)+OUT_INTEREST+1+CoinToDouble(nFees)))
                             {
-
-                                if (fDebug3) LogPrintf("ConnectBlockError[ResearchAge] : Researchers Reward Pays too much : Interest %f and Research %f and StakeReward %f, OUT_POR %f, with Out_Interest %f for CPID %s ",
-                                    (double)bb.InterestSubsidy,(double)bb.ResearchSubsidy,dStakeReward,(double)OUT_POR,(double)OUT_INTEREST,bb.cpid);
-
-                                return DoS(10,error("ConnectBlock[ResearchAge] : Researchers Reward Pays too much : Interest %f and Research %f and StakeReward %f, OUT_POR %f, with Out_Interest %f for CPID %s ",
-                                    (double)bb.InterestSubsidy,(double)bb.ResearchSubsidy,dStakeReward,(double)OUT_POR,(double)OUT_INTEREST,bb.cpid.c_str()));
+                                if( vSkipHashdStakeRewardSignCheck.count(pindex->GetBlockHash())==0 || (vSkipHashdStakeRewardSignCheck.count(pindex->GetBlockHash())!=0 && !fTestNet))
```
Comment (denravonska) [inline src/main.cpp]: These hashes only affect testnet, right? If so you can change it to ehter:
```
// Compact
if(!fTestNet || vSkipHashdStakeRewardSignCheck.count(pindex->GetBlockHash()) !=0)
```
and maybe change the set name to `badSignBlocksTestnet`.

That way we don't accidentally skip a block in prod and we don't have to do the set lookup.

File: src/main.cpp (hunk 3 above)
Comment (Quezacoatl1) [reply src/main.cpp:3414]: I never touched those lines except for copying, can you elaborate what you mean?

File: src/main.cpp (hunk 2 above)
Comment (Quezacoatl1) [reply src/main.cpp]: fixed, updating when I fully understood all your proposals

File: src/main.cpp (hunk 4 above)
Comment (Quezacoatl1) [reply src/main.cpp]: Yes, those exceptions are testnet only.

Changed the set name and compacted the if statement. Thanks!

Are you certain your if statement is correct? If I got it right (and I just copied form a few lines above within the code) I need 

vSkipHashdStakeRewardSignCheck.count(pindex->GetBlockHash()) ==0

to check that the current block is NOT amongst the exceptions.

Comment (tomasbrod) [reply src/main.cpp]: This is getting complicated. You only conditioning the error return. So you want to return error only if( (it is not testnet) or (the block is not in exception list) ).

File: src/main.cpp (hunk 3 above)
Comment (tomasbrod) [reply src/main.cpp:3414]: The diff clearly shows you did.
I can fix it for you later.

File: src/main.cpp [hunk 5]
```diff
@@ -3388,20 +3388,32 @@ bool CBlock::ConnectBlock(CTxDB& txdb, CBlockIndex* pindex, bool fJustCheck, boo
...
-
-                                return DoS(10,error("ConnectBlock[ResearchAge] : Researchers Reward Pays too much : Interest %f and Research %f and StakeReward %f, OUT_POR %f, with Out_Interest %f for CPID %s ",
-                                    (double)bb.InterestSubsidy,(double)bb.ResearchSubsidy,dStakeReward,(double)OUT_POR,(double)OUT_INTEREST,bb.cpid.c_str()));
+                                if(!fTestNet || badSignBlocksTestnet.count(pindex->GetBlockHash()) ==0)
+                                {
+                                    if (fDebug3) LogPrintf("ConnectBlockError[ResearchAge] : Researchers Reward Pays too much : Interest %f and Research %f and StakeReward %f, OUT_POR %f, with Out_Interest %f for CPID %s ",
+                                        (double)bb.InterestSubsidy,(double)bb.ResearchSubsidy,dStakeReward,(double)OUT_POR,(double)OUT_INTEREST,bb.cpid);
+                                        return DoS(10,error("ConnectBlock[ResearchAge] : Researchers Reward Pays too much : Interest %f and Research %f and StakeReward %f, OUT_POR %f, with Out_Interest %f for CPID %s ",
```
Comment (iFoggz) [inline src/main.cpp:3413]: one ident too deep :p it happens. your learning; do not give up!

File: src/main.cpp (hunk 3 above)
Comment (Quezacoatl1) [reply src/main.cpp:3414]: omg I just did not know what an indent was. sorry

**Comment by Quezacoatl1** [comment]
Guys thanks for your comments and reviews but I fear I am stealing you more time than useful because I am just not into coding. I suggest you take the hashes and quickly put those with a suitable if check and a good coding style in the code yourself ;)

**Comment by denravonska** [comment]
Replaced by #1268

**Pull request closed without merging**